File: src/api/twin/twinSetup.ts
import { invokeWithTimeout as invoke } from "@/lib/tauriInvoke";
import type { SetupOpener } from "@/lib/bindings/SetupOpener";
import type { SetupReadiness } from "@/lib/bindings/SetupReadiness";
import type { SetupSessionSnapshot } from "@/lib/bindings/SetupSessionSnapshot";
import type { SetupSteer } from "@/lib/bindings/SetupSteer";

// ============================================================================
// Twin setup plan (spark twin-setup-plan)
//
// Every call returns the WHOLE session snapshot, so the caller replaces its
// state instead of merging. None of these waits on an LLM: planning and
// reconciling run in the background and announce themselves with
// `EventName.TWIN_SETUP_UPDATED`, on which the caller refetches via
// `setupGet`. Default timeouts are therefore enough.
//
// `locale` is the app's language code (the guide asks in it); `readiness` is
// the client's reading of which slots are filled, handed to the planner.
// ============================================================================

/** The current session. Pure read. */
export const setupGet = (twinId: string) =>
  invoke<SetupSessionSnapshot>("twin_setup_get", { twinId });

/**
 * Open (or resume) the session. `opener` is a question the client can ask
 * immediately, before any plan exists.
 */
export const setupOpen = (
  twinId: string,
  readiness: SetupReadiness,
  locale?: string | null,
  opener?: SetupOpener | null,
) =>
  invoke<SetupSessionSnapshot>("twin_setup_open", {
    twinId,
    locale: locale ?? null,
    readiness,
    opener: opener ?? null,
  });

/** Answer the live step; `answer === null` skips it. */
export const setupAnswer = (
  twinId: string,
  stepId: string,
  answer: string | null,
  readiness: SetupReadiness,
  locale?: string | null,
) =>
  invoke<SetupSessionSnapshot>("twin_setup_answer", {
    twinId,
    stepId,
    answer,
    locale: locale ?? null,
    readiness,
  });

/** Steer the plan (`{ action: "dropGoal", goalId }`, `{ action: "redeal" }`, ...). */
export const setupSteer = (
  twinId: string,
  steer: SetupSteer,
  readiness: SetupReadiness,
  locale?: string | null,
) =>
  invoke<SetupSessionSnapshot>("twin_setup_steer", {
    twinId,
    steer,
    locale: locale ?? null,
    readiness,
  });

/** Record the verdict on an offer. */
export const setupOfferVerdict = (
  twinId: string,
  offerId: string,
  verdict: "accepted" | "edited" | "dismissed",
) =>
  invoke<SetupSessionSnapshot>("twin_setup_offer_verdict", {
    twinId,
    offerId,
    verdict,
  });

/** Throw the plan away and build a fresh one. */
export const setupRebuild = (
  twinId: string,
  readiness: SetupReadiness,
  locale?: string | null,
) =>
  invoke<SetupSessionSnapshot>("twin_setup_rebuild", {
    twinId,
    locale: locale ?? null,
    readiness,
  });

File: src-tauri/src/commands/infrastructure/twin_setup.rs
//! Twin setup plan — the command surface (spark `twin-setup-plan`).
//!
//! Six adapters, each returning the whole [`SetupSessionSnapshot`] so the
//! client renders one shape and never merges partial state. None of them waits
//! on an LLM: planning and reconciling run in the background and announce
//! themselves with `twin-setup-updated` (`event_name::TWIN_SETUP_UPDATED`,
//! payload `SetupUpdatedEvent`), after which the client calls
//! [`twin_setup_get`].
//!
//! Each command is a thin adapter over `engine::twin_setup` (the operations
//! run in one transaction off the IPC thread, then schedule the background
//! plan / reconcile / refill work on the per-twin worker).

use std::sync::Arc;

use tauri::{AppHandle, State};

use crate::db::models::{SetupOpener, SetupReadiness, SetupSessionSnapshot, SetupSteer};
use crate::engine::twin_setup::{self as engine, JobCtx};
use crate::error::AppError;
use crate::ipc_auth::require_auth;
use crate::AppState;

fn ctx(app: &AppHandle, state: &AppState) -> JobCtx {
    JobCtx::for_app(app, state.db.clone())
}

/// The current session. Pure read.
#[tauri::command]
pub async fn twin_setup_get(
    state: State<'_, Arc<AppState>>,
    twin_id: String,
) -> Result<SetupSessionSnapshot, AppError> {
    require_auth(&state).await?;
    engine::get(&state.db, twin_id).await
}

/// Open (or resume) the setup session: ensure a plan (the deep pass starts in
/// the background when there is none, it failed, or its lease went stale) and
/// a live step (the queue's head, else `opener`).
#[tauri::command]
pub async fn twin_setup_open(
    app: AppHandle,
    state: State<'_, Arc<AppState>>,
    twin_id: String,
    locale: Option<String>,
    readiness: SetupReadiness,
    opener: Option<SetupOpener>,
) -> Result<SetupSessionSnapshot, AppError> {
    require_auth(&state).await?;
    engine::open(&ctx(&app, &state), twin_id, locale, readiness, opener).await
}

/// Answer (or skip, with `answer == None`) the live step. A `step_id` that is
/// not the live step is a no-op returning the snapshot unchanged.
#[tauri::command]
pub async fn twin_setup_answer(
    app: AppHandle,
    state: State<'_, Arc<AppState>>,
    twin_id: String,
    step_id: String,
    answer: Option<String>,
    locale: Option<String>,
    readiness: SetupReadiness,
) -> Result<SetupSessionSnapshot, AppError> {
    require_auth(&state).await?;
    engine::answer(
        &ctx(&app, &state),
        twin_id,
        step_id,
        answer,
        locale,
        readiness,
    )
    .await
}

/// Apply an operator steer (drop/pin/restore a goal, ask a step next, change
/// stage/topic/focus, hand questions over, redeal the queue).
#[tauri::command]
pub async fn twin_setup_steer(
    app: AppHandle,
    state: State<'_, Arc<AppState>>,
    twin_id: String,
    steer: SetupSteer,
    locale: Option<String>,
    readiness: SetupReadiness,
) -> Result<SetupSessionSnapshot, AppError> {
    require_auth(&state).await?;
    engine::steer(&ctx(&app, &state), twin_id, steer, locale, readiness).await
}

/// Record the operator's verdict on an offer: `accepted` | `edited` |
/// `dismissed`. The field write itself stays client-side.
#[tauri::command]
pub async fn twin_setup_offer_verdict(
    app: AppHandle,
    state: State<'_, Arc<AppState>>,
    twin_id: String,
    offer_id: String,
    verdict: String,
) -> Result<SetupSessionSnapshot, AppError> {
    require_auth(&state).await?;
    engine::offer_verdict(&ctx(&app, &state), twin_id, offer_id, verdict).await
}

/// Build a fresh plan over everything on file. Transcript, offers and
/// observations are kept.
#[tauri::command]
pub async fn twin_setup_rebuild(
    app: AppHandle,
    state: State<'_, Arc<AppState>>,
    twin_id: String,
    locale: Option<String>,
    readiness: SetupReadiness,
) -> Result<SetupSessionSnapshot, AppError> {
    require_auth(&state).await?;
    engine::rebuild(&ctx(&app, &state), twin_id, locale, readiness).await
}
